src/snapshot_middleware/mod.rs

```diff
@@ -54,89 +63,232 @@ pub fn snapshot_from_vfs(
+                if rule.matches(&init_path) {
+                    return match rule.middleware {
+                        Middleware::Project => snapshot_project(context, vfs, &init_path),
 
-        let init_path = path.join("init.luau");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
-        }
+                        Middleware::ModuleScript => {
+                            snapshot_lua_init(context, vfs, &init_path, ScriptType::Module)
+                        }
+                        Middleware::ServerScript => {
+                            snapshot_lua_init(context, vfs, &init_path, ScriptType::Server)
+                        }
+                        Middleware::ClientScript => {
+                            snapshot_lua_init(context, vfs, &init_path, ScriptType::Client)
+                        }
 
-        let init_path = path.join("init.lua");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
-        }
+                        Middleware::Csv => snapshot_csv_init(context, vfs, &init_path),
 
-        let init_path = path.join("init.server.luau");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
+                        _ => snapshot_dir(context, vfs, path),
+                    };
+                }
+            }
+            snapshot_dir(context, vfs, path)
+        } else {
+            snapshot_dir(context, vfs, path)
         }
+    } else {
+        let file_name = path
+            .file_name()
+            .and_then(|n| n.to_str())
+            .with_context(|| format!("file name of {} is invalid", path.display()))?;
 
-        let init_path = path.join("init.server.lua");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
+        // TODO: Is this even necessary anymore?
+        match file_name {
+            "init.server.luau" | "init.server.lua" | "init.client.luau" | "init.client.lua"
+            | "init.luau" | "init.lua" | "init.csv" => return Ok(None),
+            _ => {}
         }
 
-        let init_path = path.join("init.client.luau");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
-        }
+        snapshot_from_path(context, vfs, path)
+    }
+}
 
-        let init_path = path.join("init.client.lua");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
-        }
+/// Gets an `init` path for the given directory.
+/// This uses an intrinsic priority list and for compatibility,
+/// it should not be changed.
+fn get_init_path<P: AsRef<Path>>(vfs: &Vfs, dir: P) -> anyhow::Result<Option<PathBuf>> {
```

Comment:
All of these have the same if block. We can make this an iteration over potential init names and run the same condition for each.

Reply:
The intent is we'll eventually rip this out anyway, so I'm going to decline to do it for now.